### ISSUE TITLE
Create .pyi code stubs for pyodide, pyodide.ffi, pyodide.http, etc

### ISSUE THREAD

**Issue by simonegiacomelli** [pyodide/pyodide#3672]
## 🚀 Feature

Create stub for pyodide, pyodide.ffi, pyodide.http, etc

### Motivation

I often find myself wanting for ide completion and automatic assisted imports. 
Right now the ide underlines pyodide imports as errors.

### Pitch

I already did pyi both for pyodide and for the web-dom, so I have some experienge. 
I would be happy to contribute with some pyodide pyi infrastructure.

### Additional context

I'm building a library that facilitates the use of pyodide and the communication between python/pyodide and python/server so I often use pyodide.ffi and related imports.

**Comment by rth** [comment]
We do have type annotations for those modules already, so shouldn't it be just a matter of exposing them to IDE: maybe we need add the `py.typed` file?

But yes, if you want to work on making the situation better that would be certainly very welcome!

Generating Python types for JS objects accessed via the API, would be a more ambitious project, the inverse of https://github.com/pyodide/pyodide/issues/2162

**Comment by hoodmane** [comment]
I think all you need to do is `pip install pyodide_py` and you get type annotations. The main problem is that this feature is completely undocumented.

**Comment by hoodmane** [comment]
I think all you need to do is `pip install pyodide_py` and you get type annotations. The main problem is that this feature is completely undocumented.

**Comment by hoodmane** [comment]
Though also I'm not sure if `js.pyi` is included in `pyodide-py`. It ought to be...

**Comment by hoodmane** [comment]
> I would be happy to contribute with some pyodide pyi infrastructure.

Thanks! We don't need pyi files for the pyodide package itself since we already have types for it. But if you are interested in adding more apis to `js.pyi` that would be fantastic.

**Comment by simonegiacomelli** [comment]
> I think all you need to do is `pip install pyodide-py` and you get type annotations. The main problem is that this feature is completely undocumented.

@hoodmane Thanks! I didn't know the existance of `pyodide-py`.

I tried the package and I noticed that some imports, e.g., create_proxy refers to `_pyodide._core_docs`.
Looking through the files I saw only .py and no .pyi. 

Is this package needed for python execution somewhere?

**Comment by hoodmane** [comment]
Yeah the file`js.pyi` should be in there but it isn't currently. You can find it here: https://github.com/pyodide/pyodide/blob/main/src/py/js.pyi

**Comment by simonegiacomelli** [comment]
> Thanks! We don't need pyi files for the pyodide package itself since we already have types for it. But if you are interested in adding more apis to `js.pyi` that would be fantastic.

I don't know how to find your `js.pyi`...; could you point me to their location?

In the last months I developed a [generator](https://github.com/simonegiacomelli/pynanto2/tree/main/js_pyi) that takes in .webidl files and generates a js.pyi.
It works quite well and I covered many cases and imported as many definitions as possible (HTMLElement hierarchy, bluetooth api, webgl, etc.).
You can see the last generated [js.pyi here.](https://github.com/simonegiacomelli/pynanto2/blob/main/src/js/__init__.pyi)

Having the IDE completion is a life saver when developing with pyodide.

I still have a [few things](https://github.com/simonegiacomelli/pynanto2/issues?q=is%3Aissue+is%3Aopen++js.pyi) that I would like to do, for example adding the comments or link directly the MDN webpage.


Let me know if this could be of interest; I would be glad to make it available to everybody.

**Comment by simonegiacomelli** [comment]
@hoodmane if you want to try the js.pyi you can `pip install pynanto`.

**Comment by hoodmane** [comment]
Yes this looks very nice thanks @simonegiacomelli we would definitely be interested in this.

One thing that's useful in our (much more limited hand written) `js.pyi` file is the following code:
```py
from _pyodide._core_docs import _JsProxyMetaClass

class _JsMeta(_JsProxyMetaClass, JsProxy):
    pass

class _JsObject(metaclass=_JsMeta):
    pass
```
Then subclasses of `_JsObject` are considered to be instances of `JsProxy` (the classes themselves are instances of `JsProxy`).

**Comment by simonegiacomelli** [comment]
@hoodmane would you like to add `_JsMeta` the the generated js.pyi? 

Right now the .pyi created with the generator is integrated with my pynanto/wwwpy repo... I think this could be a separate pypi package so people can install it separately. I already thought to do it but I didn't to save resource and lower the dependencies of pynanto/wwwpy.

What do you think?

**Comment by simonegiacomelli** [comment]
@hoodmane this is a clarification request about two issues, this one and #3470.

I have the preconception that these needs to be two different issues because they seems to me different features
- #3672 (this one) is about pyodide stubs, e.g., create_proxy, pyfetch, to_js. And maybe more: pyodide_js.loadPackage, micropip.install. In other words it deals with pyodide specific interfaces.
- #3470 deals with the dom, in its various declinations: browser, nodejs, worker (sorry if I misname them). In other words it needs to understand the w3c Web specs and webidls (see more details in that issue).

Where I already have considerable work is in the js.pyi part. I may contribute with this issue in the future but not right now. I opened this one in the hope something was already present.

So, probably some of my previous replies in this issue are kind of misplaced and should go in the other one.
If you agree with this reasoning the other issue should be reopened, and after agreeing on how to proceed I may be able to share my webidl parser and js.pyi generator.

What do you think?

**Comment by hoodmane** [comment]
So I think the only issue there is it isn't well documented. We should add an faq or possibly a quick mention of types in the API docs.

**Comment by ryanking13** [comment]
> Maybe we could move it into pyodide/web-idl or something like that if you want. I'm also curious whether @ryanking13 or @rth think.

Sounds like a lovely idea to me. I think @simonegiacomelli can go ahead and create a webidl parser in a separate repository, and we can use it in pyodide. We can put the repository to Pyodide organization if you are willing to.

**Comment by simonegiacomelli** [comment]
@ryanking13 @hoodmane 
Sounds good to me. I think using `pyodide/web-idl` is a good idea.
The name `web-idl` is also free on pypi.

> And what about the concept of making synonyms for js for different realms? It's not really ideal ...
I have an idea that is not a solution to this but that could open the subject for further development.
One of the first steps to make this possible is gather the information and make it available and easy to use (what api is available where).
For starters, this information could end up in the api comments; it could also partially fix those hybrid situations (e.g., the Web Bluetooth api being available only in certain browsers/platforms)

I already thought to add comments in some way; after the [last iteration](https://github.com/simonegiacomelli/pynanto2/issues/8) of research on this topic I think the best source could be the MDN for comments and webidl for the platform availability.
If you know other/better resources, let me know.

If it's ok with you, we could move this discussion and related to the new repository.

What do you think?

**Comment by hoodmane** [comment]
Created an empty `pyodide/web-idl` repo and invited @simonegiacomelli with write access.

**Comment by simonegiacomelli** [comment]
Created an [issue](https://github.com/pyodide/web-idl/issues/1) to continue the discussion about `pyodide/web-idl` python stubs
